### PR TITLE
Reconfigure logging to follow best practices and catch exceptions in generator thread

### LINKED ISSUE #36: Exceptions on generator thread not being logged
**Describe the bug**
Exceptions related to the main app window seem to be logged just fine, but exceptions in the generator loop (which is run on a separate thread) are not logged properly.

**To Reproduce**
1. Add an intentional error to the generator loop
2. Run program until exception is thrown

**Expected behavior**
The exception should be logged.

**Screenshots**
N/A

**Additional context**
N/A

### LINKED ISSUE #43: Reconfigure logging according to Python best practices
**Is your feature request related to a problem? Please describe.**
Currently the application logs only to the root logger, which will cause visibility issues if any of the core modules are exported as a library, and doesn't allow a developer to set module-specific logging levels.

**Describe the solution you'd like**
Replace the calls to the root logger with a logger object per module. This better adheres to [Python best practices](https://docs.python.org/3/howto/logging.html#when-to-use-logging), and allows a lot more flexibility for the developer/downstream consumers of the core modules when configuring logging.

**Describe alternatives you've considered**
N/A

**Additional context**
Related to #36 since it's a rewrite of logging.

### PR DESCRIPTION
**Pull request by ethanorr**
# Description
Updates the logging configuration/usage to more closely follow current Python best practices. Configuration is now done through a json file, and a logger object is created per module which can be individually configured through the configuration file. This file is read at runtime and shipped with the executable, so the end user can modify the configuration file in order to capture more specific logs (which may be helpful for us to capture debug information if a non-technical user encounters a problem unique to their hardware).

Fixes #43

Additionally, these changes allow us to override the excepthook for the generator thread with one that logs to our file handler rather than syserr (the default behavior).

Fixes #36 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- Build and run the application without connecting to an iRacing instance, and observe that each module logs to the generated log file on startup.

Previous startup sequence logs:
```
2024-12-04 18:54:17,756 (main.py) [INFO] Program started
2024-12-04 18:54:17,756 (app.py) [INFO] Initializing main application window
2024-12-04 18:54:17,822 (app.py) [INFO] Loading tooltips text
2024-12-04 18:54:17,830 (app.py) [INFO] Loading settings from settings.ini
2024-12-04 18:54:17,836 (generator.py) [INFO] Initializing safety car generator
2024-12-04 18:54:17,837 (app.py) [INFO] Creating widgets for main application window
2024-12-04 20:48:48,408 (app.py) [INFO] Closing main application window
```
Observed startup sequence logs:
```
2024-12-04 21:15:47,923 (__main__) [INFO] Program started
2024-12-04 21:15:47,923 (core.app) [INFO] Initializing main application window
2024-12-04 21:15:47,990 (core.app) [INFO] Loading tooltips text
2024-12-04 21:15:47,997 (core.app) [INFO] Loading settings from settings.ini
2024-12-04 21:15:47,998 (core.generator) [INFO] Initializing safety car generator
2024-12-04 21:15:47,998 (core.app) [INFO] Creating widgets for main application window
2024-12-04 21:17:34,880 (core.app) [INFO] Saving settings to config file
2024-12-04 21:17:34,880 (core.generator) [INFO] Connecting to iRacing
2024-12-04 21:18:33,989 (core.app) [INFO] Closing main application window
```

 - Additionally, I threw an exception in the generator thread (you have to start an AI session in order to get the generator thread to run), and verified that the uncaught exception is logged as expected with traceback information.
```
2024-12-05 16:37:27,719 (__main__) [INFO] Program started
2024-12-05 16:37:27,719 (core.app) [INFO] Initializing main application window
2024-12-05 16:37:27,786 (core.app) [INFO] Loading tooltips text
2024-12-05 16:37:27,793 (core.app) [INFO] Loading settings from settings.ini
2024-12-05 16:37:27,795 (core.generator) [INFO] Initializing safety car generator
2024-12-05 16:37:27,795 (core.app) [INFO] Creating widgets for main application window
2024-12-05 16:38:42,678 (core.app) [INFO] Saving settings to config file
2024-12-05 16:38:42,679 (core.generator) [INFO] Connecting to iRacing
2024-12-05 16:38:44,173 (core.generator) [CRITICAL] Uncaught exception:
Traceback (most recent call last):
  File "threading.py", line 1041, in _bootstrap_inner
  File "threading.py", line 992, in run
  File "core\generator.py", line 235, in _loop
Exception: Test logging from generator thread
2024-12-05 16:40:09,102 (core.app) [INFO] Closing main application window
```

### CODE REVIEW

**Comment by ethanorr** [comment]
Updated the PR to add a quick fix for #36 since it relies on these changes -- @joshjaysalazar please let me know if you'd rather I revert that commit and make another branch/PR for that issue after this one goes through, or if combining these two issues into one PR (as I've done here) is okay.

**Comment by joshjaysalazar** [comment]
> Updated the PR to add a quick fix for #36 since it relies on these changes -- @joshjaysalazar please let me know if you'd rather I revert that commit and make another branch/PR for that issue after this one goes through, or if combining these two issues into one PR (as I've done here) is okay.

I think it's fine to do them both on this PR, since they're related. I've definitely done that for previous PRs. The branch I'm working on now will be solving 2 issues, in fact.

File: src/build.bat [hunk 1]
```diff
@@ -9,6 +9,8 @@ pyinstaller --noconfirm --log-level=FATAL --noconsole --onefile --hidden-import=
 ECHO Copying files to dist...
 :: Copy settings.ini to dist
 COPY settings.ini dist\settings.ini
+:: Copy logging.yaml to dist
```
Comment (joshjaysalazar) [inline src/build.bat:12]: Comment refers to a yaml, but the file is a JSON file.

File: src/main.py [hunk 2]
```diff
@@ -11,19 +13,21 @@ def setup_logging():
...
+    logger = logging.getLogger(__name__)
+
+    # Configure logging
+    with open("logging.json") as logging_conf_file:
+        logging_conf = json.load(logging_conf_file)
+
+    # Dynamically set log file name to current time
     current_datetime = datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
+    logfile = logging_conf['handlers']['file']['filename']
+    logging_conf['handlers']['file']['filename'] = logfile.replace("{current_datetime}", current_datetime)
```
Comment (joshjaysalazar) [inline src/main.py:25]: Could you change the quotes on 24 and 25 to double quotes just for consistency?

**Review by joshjaysalazar — CHANGES_REQUESTED** [review]
Just a couple of very small changes for style consistency, but it all looks good overall to me. I clearly need to read up more on the capabilities of the logging library! Thanks for correcting this.

File: src/main.py (hunk 2 above)
Comment (ljtijhuis) [reply src/main.py:25]: https://github.com/astral-sh/ruff 👀

File: src/build.bat (hunk 1 above)
Comment (ethanorr) [reply src/build.bat]: Good catch! Resolved in da65d5f.

File: src/main.py (hunk 2 above)
Comment (ethanorr) [reply src/main.py]: Agreed, fixed in ed271c4

**Review by joshjaysalazar — APPROVED** [review]
Looks good, thank you!

**Pull request merged**